python/cuml/neighbors/nearest_neighbors.pyx

```diff
@@ -393,5 +431,174 @@ class NearestNeighbors(Base):
+                             ' or "distance" but got "%s" instead' % mode)
+
+        indices = ind_mlarr.to_output('cupy')[:, 1:] if X is None \
+            else ind_mlarr.to_output('cupy')
+        n_samples = indices.shape[0]
+        n_samples_fit = self.X_m.shape[0]
+        n_nonzero = n_samples * n_neighbors
+        rowptr = cp.arange(0, n_nonzero + 1, n_neighbors)
+
+        sparse_csr = cp.sparse.csr_matrix((distances, cp.ravel(indices),
+                                          rowptr), shape=(n_samples,
+                                          n_samples_fit))
+
+        if self._get_output_type(X) is 'numpy':
+            return sparse_csr.get()
+        else:
+            return sparse_csr
+
+
+def kneighbors_graph(X=None, n_neighbors=5, mode='connectivity', verbose=False,
+                     handle=None, algorithm="brute", metric="euclidean", p=2,
+                     include_self=False, metric_params=None, output_type=None):
+    """
+    Computes the (weighted) graph of k-Neighbors for points in X.
+
+    Parameters
+    ----------
+    X : array-like (device or host) shape = (n_samples, n_features)
+        Dense matrix (floats or doubles) of shape (n_samples, n_features).
+        Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+        ndarray, cuda array interface compliant array like CuPy
+
+    n_neighbors : Integer
+        Number of neighbors to search. If not provided, the n_neighbors
+        from the model instance is used (default=5)
+
+    mode : string (default='connectivity')
+        Values in connectivity matrix: 'connectivity' returns the
+        connectivity matrix with ones and zeros, 'distance' returns the
+        edges as the distances between points with the requested metric.
+
+    verbose : int or boolean (default = False)
+        Logging level
+
+    handle : cumlHandle
+        The cumlHandle resources to use
+
+    algorithm : string (default='brute')
+        The query algorithm to use. Currently, only 'brute' is supported.
+
+    metric : string (default='euclidean').
+        Distance metric to use. Supported distances are ['l1, 'cityblock',
+        'taxicab', 'manhattan', 'euclidean', 'l2', 'braycurtis', 'canberra',
+        'minkowski', 'chebyshev', 'jensenshannon', 'cosine', 'correlation']
+
+    p : float (default=2) Parameter for the Minkowski metric. When p = 1, this
+        is equivalent to manhattan distance (l1), and euclidean distance (l2)
+        for p = 2. For arbitrary p, minkowski distance (lp) is used.
+
+    include_self : bool or 'auto' (default=False)
+        Whether or not to mark each sample as the first nearest neighbor to
+        itself. If 'auto', then True is used for mode='connectivity' and False
+        for mode='distance'.
+
+    metric_params : dict, optional (default = None) This is currently ignored.
+
+    output_type : {'input', 'cudf', 'pandas', 'cupy', 'numpy'}, optional
```

Comment:
Should probably keep only `input`, `cupy`, `numpy` until the official decision is mad to support Dataframes as output.

Reply:
My bad must have missed that one, fixed it.